sympy/combinatorics/homomorphisms.py

```diff
@@ -417,3 +419,126 @@ def block_homomorphism(group, blocks):
+    gens = G.generators
+    for subset in itertools.permutations(H, len(gens)):
+        # Create a map
+        func_map = dict(zip(gens, subset))
+        counterexample = False
+        # Loop through the mapped elements and try to extend the mapping
+        # or to find a counter example.
+        while not counterexample:
+            extended_map = {}
+            for g, h in itertools.product(func_map, func_map):
+                if g*h in func_map:
+                    if func_map[g]*func_map[h] != func_map[g*h]:
+                        counterexample = True
+                        break
+                else:
+                    extended_map[g*h] = func_map[g]*func_map[h]
+
+            # Break when all the elements are mapped.
+            if len(func_map) == H.order() :
+                break
+
+            # Remove the duplicate elements in extended_map
+            # and merge the extended_map with the map.
+            image_len = len(set(func_map.values()) | set(extended_map.values()))
+            if image_len != len(func_map) + len(extended_map):
+                counterexample = True
+                break
+            func_map.update(extended_map)
+
+        if not counterexample:
+            # Returns a `GroupHomomorphism` object.
+            return homomorphism(G, H, G.generators, subset, check=False)
+
+    return None
+
+def is_isomorphic(G, H, isomorphism=False):
+    '''
+    Check if the given groups are isomorphic.
+
+    Arguments:
+        G (a finite `FpGroup` or a `PermutationGroup`) -- First group
+        H (a finite `FpGroup` or a `PermutationGroup`) -- Second group
+        isomorphism(boolean) -- This computes an isomorphism when set to True
+
+    Returns:
+    if isomorphism is set to `False` --
+        `True` if the groups are isomorphic
+        `False` if the groups are not isomorphic
+    if isomorphism is set to `True` --
+        Returns a tuple (boolean, `GroupHomomorphism` object).
+
+    Examples
+    ========
+
+    >>> from sympy.combinatorics import Permutation
+    >>> from sympy.combinatorics.perm_groups import PermutationGroup
+    >>> from sympy.combinatorics.homomorphisms import homomorphism, group_isomorphism, is_isomorphic
+    >>> from sympy.combinatorics.free_groups import free_group
+    >>> from sympy.combinatorics.fp_groups import FpGroup
+    >>> from sympy.combinatorics.named_groups import DihedralGroup, AlternatingGroup
+
+    >>> D = DihedralGroup(8)
+    >>> p = Permutation(0, 1, 2, 3, 4, 5, 6, 7)
+    >>> P = PermutationGroup(p)
+    >>> is_isomorphic(D, P, isomorphism=True)
+    (False, None)
+
+    >>> F, a, b = free_group("a, b")
+    >>> G = FpGroup(F, [a**3, b**3, (a*b)**2])
+    >>> H = AlternatingGroup(4)
+    >>> G = G._to_perm_group()[0]
+    >>> (check, T) = is_isomorphic(G, H, isomorphism=True)
+    >>> check
+    True
+    >>> T(Permutation(0, 9, 6)(1, 10, 8)(2, 11, 4)(3, 5, 7))
+    (0 2 3)
+
+    '''
+    T = group_isomorphism(G, H)
```

Comment:
If `isomorphism = False`, before we try this, we should try to see if the orders of `G` and `H` are equal to the same cyclic number `n`, i.e. something that satisfies `gcd(n, φ(n)) = 1` (where `φ` is Euler's totient) - then we know both of them are cyclic so can return `True`.

Reply:
That should be `G.order()` in sieve.